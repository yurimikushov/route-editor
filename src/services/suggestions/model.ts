type Suggestions = {
  isLoading: boolean
  suggestions: Array<Suggestion>
  error: Error | null
  load: (address: string) => Promise<void>
  clear: () => void
}

type SuggestionsState = {
  isLoading: boolean
  suggestions: Array<Suggestion>
  error: Error | null
}

type Suggestion = {
  name: string
  description: string
  point: { lat: number; lon: number }
}

export type { Suggestions, SuggestionsState, Suggestion }
