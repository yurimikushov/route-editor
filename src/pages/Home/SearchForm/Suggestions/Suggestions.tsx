import { FC, useState } from 'react'
import cn from 'classnames'
import size from 'lodash/size'
import map from 'lodash/map'
import useKeyDown from 'hooks/useKeyDown'
import Suggestion from './Suggestion'
import SuggestionsProps from './Suggestions.props'

const Suggestions: FC<SuggestionsProps> = ({
  className,
  suggestions,
  onChoose,
}) => {
  const [highlightedSuggestion, setHighlightedSuggestion] = useState(-1)

  useKeyDown(
    'ArrowUp',
    () => {
      setHighlightedSuggestion((highlightedSuggestion) => {
        if (highlightedSuggestion === -1) {
          return size(suggestions) - 1
        }

        return highlightedSuggestion - 1
      })
    },
    [suggestions]
  )

  useKeyDown(
    'ArrowDown',
    () => {
      setHighlightedSuggestion((highlightedSuggestion) => {
        if (highlightedSuggestion === size(suggestions) - 1) {
          return 0
        }

        return highlightedSuggestion + 1
      })
    },
    [suggestions]
  )

  useKeyDown(
    'Enter',
    () => {
      if (highlightedSuggestion === -1) {
        return
      }

      onChoose(suggestions[highlightedSuggestion])
    },
    [highlightedSuggestion, suggestions]
  )

  return (
    <ul className={cn(className, 'py-1 w-full rounded-md shadow-lg')}>
      {map(suggestions, (suggestion, i) => (
        <Suggestion
          key={`${suggestion.point.lat}::${suggestion.point.lon}`}
          isHighlighted={highlightedSuggestion === i}
          suggestion={suggestion}
          onChoose={() => onChoose(suggestion)}
        />
      ))}
    </ul>
  )
}

export default Suggestions
