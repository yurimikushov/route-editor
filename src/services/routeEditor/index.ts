export { default as RouteEditorProvider } from './RouteEditorProvider'
export { useRouteEditor } from './RouteEditorContext'
export type { RouteEditor, Address, Point } from './model'
