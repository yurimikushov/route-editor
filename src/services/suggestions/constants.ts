export const LOAD_BEGIN = 'suggestions/LOAD_BEGIN'
export const LOAD_SUCCESS = 'suggestions/LOAD_SUCCESS'
export const LOAD_FAIL = 'suggestions/LOAD_FAIL'
export const CLEAR = 'suggestions/CLEAR'
