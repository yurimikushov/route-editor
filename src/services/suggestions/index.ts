export { default as SuggestionsProvider } from './SuggestionsProvider'
export { useSuggestions } from './SuggestionsContext'
export type { Suggestion } from './model'
