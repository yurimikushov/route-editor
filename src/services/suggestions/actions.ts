import { LOAD_BEGIN, LOAD_SUCCESS, LOAD_FAIL, CLEAR } from './constants'
import { Suggestion } from './model'

const loadBegin = () =>
  ({
    type: LOAD_BEGIN,
  } as const)

const loadSuccess = (suggestions: Array<Suggestion>) =>
  ({
    type: LOAD_SUCCESS,
    payload: {
      suggestions,
    },
  } as const)

const loadFail = (error: Error) =>
  ({
    type: LOAD_FAIL,
    payload: {
      error,
    },
  } as const)

const clear = () =>
  ({
    type: CLEAR,
  } as const)

type LoadBegin = ReturnType<typeof loadBegin>
type LoadSuccess = ReturnType<typeof loadSuccess>
type LoadFail = ReturnType<typeof loadFail>
type Clear = ReturnType<typeof clear>

type Action = LoadBegin | LoadSuccess | LoadFail | Clear

export { loadBegin, loadSuccess, loadFail, clear }
export type { Action }
