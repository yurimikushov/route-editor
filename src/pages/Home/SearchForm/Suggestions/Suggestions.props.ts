import { Suggestion } from 'services/suggestions'

type SuggestionsProps = {
  className?: string
  suggestions: Array<Suggestion>
  onSelect: (suggestion: Suggestion) => void
}

export default SuggestionsProps
