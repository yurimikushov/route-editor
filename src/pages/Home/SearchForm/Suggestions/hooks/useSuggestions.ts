import { useState } from 'react'
import size from 'lodash/size'
import { Suggestion } from 'services/suggestions'
import useKeyDown from 'hooks/useKeyDown'

const useSuggestions = (
  suggestions: Array<Suggestion>,
  onChoose: (suggestion: Suggestion) => void
) => {
  const [highlightedSuggestion, setHighlightedSuggestion] = useState(-1)

  useKeyDown(
    'ArrowUp',
    () => {
      setHighlightedSuggestion((highlightedSuggestion) => {
        if (highlightedSuggestion === -1) {
          return size(suggestions) - 1
        }

        return highlightedSuggestion - 1
      })
    },
    [suggestions]
  )

  useKeyDown(
    'ArrowDown',
    () => {
      setHighlightedSuggestion((highlightedSuggestion) => {
        if (highlightedSuggestion === size(suggestions) - 1) {
          return 0
        }

        return highlightedSuggestion + 1
      })
    },
    [suggestions]
  )

  useKeyDown(
    'Enter',
    () => {
      if (highlightedSuggestion === -1) {
        return
      }

      onChoose(suggestions[highlightedSuggestion])
    },
    [highlightedSuggestion, suggestions]
  )

  useKeyDown(
    'Space',
    (e: KeyboardEvent) => {
      if (highlightedSuggestion === -1) {
        return
      }

      e.preventDefault()
      onChoose(suggestions[highlightedSuggestion])
    },
    [highlightedSuggestion, suggestions]
  )

  return { highlightedSuggestion }
}

export default useSuggestions
