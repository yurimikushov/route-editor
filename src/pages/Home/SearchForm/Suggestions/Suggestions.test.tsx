import { fireEvent, render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import noop from 'lodash/noop'
import { Suggestion } from 'services/suggestions'
import useSuggestions from './hooks/useSuggestions'
import SuggestionProps from './Suggestion/Suggestion.props'
import Suggestions from './Suggestions'

// eslint-disable-next-line react/display-name
jest.mock('./Suggestion', () => (props: SuggestionProps) => (
  <li {...props} data-testId={props.suggestion.name}>
    Suggestion
  </li>
))

jest.mock('./hooks/useSuggestions')

const runMockUseSuggestions = (result: ReturnType<typeof useSuggestions>) => {
  // @ts-expect-error bad typing
  useSuggestions.mockImplementation(() => result)
}

it(`should render suggestions when suggestions isn't empty`, () => {
  runMockUseSuggestions({
    highlightedSuggestion: -1,
  })

  const tree = renderer
    .create(
      <Suggestions
        className='some-class-name'
        suggestions={[
          {
            name: 'Balalyaka',
            description: 'Some description',
            point: {
              lat: 0,
              lon: 0,
            },
          },
        ]}
        onChoose={noop}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('should render empty container when suggestions is empty', () => {
  runMockUseSuggestions({
    highlightedSuggestion: -1,
  })

  const tree = renderer
    .create(
      <Suggestions
        className='some-class-name'
        suggestions={[]}
        onChoose={noop}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it(`should highlight suggestion when useSuggestions return positive index`, () => {
  runMockUseSuggestions({
    highlightedSuggestion: 1,
  })

  const tree = renderer
    .create(
      <Suggestions
        className='some-class-name'
        suggestions={[
          {
            name: 'Balalyaka',
            description: 'Some description',
            point: {
              lat: 0,
              lon: 0,
            },
          },
          {
            name: 'Vodka',
            description: 'Some description',
            point: {
              lat: 1,
              lon: 2,
            },
          },
          {
            name: 'Pelmeny',
            description: 'Some description',
            point: {
              lat: 2,
              lon: 2,
            },
          },
        ]}
        onChoose={noop}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('should call onChoose when suggestion is chosen', () => {
  runMockUseSuggestions({
    highlightedSuggestion: -1,
  })

  let result: Suggestion | null = null
  const handleChoose = jest.fn((suggestion: Suggestion) => {
    result = suggestion
  })

  const mockSuggestion = {
    name: 'Balalyaka',
    description: 'Some description',
    point: {
      lat: 0,
      lon: 0,
    },
  }

  render(
    <Suggestions
      className='some-class-name'
      suggestions={[mockSuggestion]}
      onChoose={handleChoose}
    />
  )

  fireEvent.mouseDown(screen.getByTestId('Balalyaka'))

  expect(result).toBe(mockSuggestion)
  expect(handleChoose).toHaveBeenCalledTimes(1)
})
