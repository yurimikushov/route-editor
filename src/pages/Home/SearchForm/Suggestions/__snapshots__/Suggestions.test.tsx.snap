// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should highlight suggestion when useSuggestions return positive index 1`] = `
<ul
  className="some-class-name py-1 w-full max-h-[30vh] md:max-h-[50vh] overflow-y-scroll md:overflow-y-auto bg-white rounded-md border border-gray-200 shadow-lg"
>
  <li
    data-testId="Balalyaka"
    isHighlighted={false}
    onMouseDown={[Function]}
    suggestion={
      Object {
        "description": "Some description",
        "name": "Balalyaka",
        "point": Object {
          "lat": 0,
          "lon": 0,
        },
      }
    }
  >
    Suggestion
  </li>
  <li
    data-testId="Vodka"
    isHighlighted={true}
    onMouseDown={[Function]}
    suggestion={
      Object {
        "description": "Some description",
        "name": "Vodka",
        "point": Object {
          "lat": 1,
          "lon": 2,
        },
      }
    }
  >
    Suggestion
  </li>
  <li
    data-testId="Pelmeny"
    isHighlighted={false}
    onMouseDown={[Function]}
    suggestion={
      Object {
        "description": "Some description",
        "name": "Pelmeny",
        "point": Object {
          "lat": 2,
          "lon": 2,
        },
      }
    }
  >
    Suggestion
  </li>
</ul>
`;

exports[`should render empty container when suggestions is empty 1`] = `
<ul
  className="some-class-name py-1 w-full max-h-[30vh] md:max-h-[50vh] overflow-y-scroll md:overflow-y-auto bg-white rounded-md border border-gray-200 shadow-lg"
/>
`;

exports[`should render suggestions when suggestions isn't empty 1`] = `
<ul
  className="some-class-name py-1 w-full max-h-[30vh] md:max-h-[50vh] overflow-y-scroll md:overflow-y-auto bg-white rounded-md border border-gray-200 shadow-lg"
>
  <li
    data-testId="Balalyaka"
    isHighlighted={false}
    onMouseDown={[Function]}
    suggestion={
      Object {
        "description": "Some description",
        "name": "Balalyaka",
        "point": Object {
          "lat": 0,
          "lon": 0,
        },
      }
    }
  >
    Suggestion
  </li>
</ul>
`;
